src/licenses/LicenseModel.cpp

```diff
@@ -0,0 +1,184 @@
+
+    LicenseItem *childItem = parentItem->child(row);
+    if (childItem)
+      return createIndex(row, column, childItem);
+    return QModelIndex();
+  }
+
+  QModelIndex LicenseModel::parent(const QModelIndex &index) const {
+    if (!index.isValid())
+      return QModelIndex();
+
+    auto *childItem = static_cast<LicenseItem *>(index.internalPointer());
+    LicenseItem *parentItem = childItem->parentItem();
+
+    if (parentItem == rootItem.get())
+      return QModelIndex();
+
+    return createIndex(parentItem->row(), 0, parentItem);
+  }
+
+  int LicenseModel::rowCount(const QModelIndex &parent) const {
+    LicenseItem *parentItem;
+    if (parent.column() > 0)
+      return 0;
+
+    if (!parent.isValid())
+      parentItem = rootItem.get();
+    else
+      parentItem = static_cast<LicenseItem *>(parent.internalPointer());
+
+    return parentItem->childCount();
+  }
+
+  int LicenseModel::columnCount(const QModelIndex &parent) const {
+    if (parent.isValid())
+      return static_cast<LicenseItem *>(parent.internalPointer())->columnCount();
+    return rootItem->columnCount();
+  }
+
+  QVariant LicenseModel::data(const QModelIndex &index, int role) const {
+    if (!index.isValid())
+      return QVariant();
+
+    if (role != licenses::LicenseFileName && role != licenses::LicenseFilePath && role != licenses::LicenseCategories)
+      return QVariant();
+
+    auto *item = static_cast<LicenseItem *>(index.internalPointer());
+
+    return item->data(role);
+  }
+
+  Qt::ItemFlags LicenseModel::flags(const QModelIndex &index) const {
+    if (!index.isValid())
+      return Qt::NoItemFlags;
+
+    return QAbstractItemModel::flags(index);
+  }
+
+  QVariant LicenseModel::headerData(int section, Qt::Orientation orientation,
+                                    int role) const {
+    if (orientation == Qt::Horizontal && role == Qt::DisplayRole)
+      return rootItem->data(section);
+
+    return QVariant();
+  }
+
+  LicenseItem::LicenseItem(QVector<QVariant> data, QList<QVariant> categories, LicenseItem *parent)
+      : m_itemData(std::move(data)), m_categories(std::move(categories)), m_parentItem(parent) {}
+
+  LicenseItem::~LicenseItem() {
+    qDeleteAll(m_childItems);
+  }
+
+  void LicenseItem::appendChild(LicenseItem *item) {
+    m_childItems.append(item);
+  }
+
+  LicenseItem *LicenseItem::child(int row) {
+    if (row < 0 || row >= m_childItems.size())
+      return nullptr;
+    return m_childItems.at(row);
+  }
+
+  int LicenseItem::childCount() const {
+    return m_childItems.count();
+  }
+
+  int LicenseItem::row() const {
+    if (m_parentItem)
+      return m_parentItem->m_childItems.indexOf(const_cast<LicenseItem *>(this));
+
+    return 0;
+  }
+
+  int LicenseItem::columnCount() const {
+    return m_itemData.count();
+  }
+
+  QVariant LicenseItem::data(int role) const {
+    switch (role) {
+      case licenses::LicenseFileName :
+        return m_itemData.at(0);
```

Comment:
might it be empty somehow or is it never so?
Here could be an out of bound exception be triggered

Reply:
Yeah, I didn't like this bit either, I rewrote it and made them real members instead...